aptos-move/aptos-release-builder/src/simulate.rs

```diff
@@ -0,0 +1,500 @@
+
+//! This module implements the simulation of governance proposals.
+//! Currently, it supports only multi-step proposals.
+//!
+//! It utilizes the the remote debugger infrastructure to fetch real chain states
+//! for local simulation, but adds another in-memory database to store the new side effects
+//! generated by the governance scripts.
+//!
+//! Normally, governance scripts needs to be approved through on-chain governance
+//! before they could be executed. This process involves setting up various states
+//! (e.g., staking pool, delegated voter), which can be quite complex.
+//!
+//! This simulation bypasses these challenges by patching specific Move functions
+//! with mock versions, most notably `fun resolve_multi_step_proposal`, thus allowing
+//! the governance process to be skipped altogether.
+//!
+//! In other words, this simulation is intended for checking whether a governance
+//! proposal will execute successfully, assuming it gets approved.
+
+use crate::aptos_framework_path;
+use anyhow::{anyhow, bail, Result};
+use aptos::{
+    common::types::PromptOptions, governance::compile_in_temp_dir, move_tool::FrameworkPackageArgs,
+};
+use aptos_gas_schedule::{AptosGasParameters, FromOnChainGasSchedule, ToOnChainGasSchedule};
+use aptos_language_e2e_tests::account::AccountData;
+use aptos_move_debugger::aptos_debugger::AptosDebugger;
+use aptos_rest_client::Client;
+use aptos_types::{
+    account_address::AccountAddress,
+    account_config::ChainIdResource,
+    on_chain_config::{GasScheduleV2, OnChainConfig},
+    state_store::{
+        in_memory_state_view::InMemoryStateView, state_key::StateKey,
+        state_storage_usage::StateStorageUsage, state_value::StateValue,
+        Result as StateStoreResult, StateView, TStateView,
+    },
+    transaction::{ExecutionStatus, Script, TransactionArgument, TransactionStatus},
+    write_set::{TransactionWrite, WriteSet},
+};
+use aptos_vm::{data_cache::AsMoveResolver, AptosVM};
+use aptos_vm_logging::log_schema::AdapterLogSchema;
+use clap::Parser;
+use move_binary_format::{
+    access::ModuleAccess,
+    file_format::{
+        AddressIdentifierIndex, Bytecode, FunctionDefinition, FunctionHandle, FunctionHandleIndex,
+        IdentifierIndex, ModuleHandle, ModuleHandleIndex, Signature, SignatureIndex,
+        SignatureToken, Visibility,
+    },
+    CompiledModule,
+};
+use move_core_types::{
+    identifier::{IdentStr, Identifier},
+    language_storage::{ModuleId, StructTag},
+    move_resource::MoveResource,
+    resolver::ModuleResolver,
+};
+use once_cell::sync::Lazy;
+use std::{
+    collections::HashMap,
+    io::Write,
+    path::{Path, PathBuf},
+    sync::Mutex,
+};
+use url::Url;
+
+/***************************************************************************************************
+ * Compiled Module Helpers
+ *
+ **************************************************************************************************/
+fn find_function_def_by_name<'a>(
```

Comment:
I am wondering if these kinds of helpers can be useful even in `third_party`?

Reply:
They could be. Although I'm thinking that maybe we should refine the APIs a little bit if we want to offer them as a general solution. Might also be a good idea to see @runtian-zhou's work on the intent engine as it ships with a script builder, and see if we can make some good unified helpers out of both use cases.

Reply:
Yea I actually have to implement those helpers in my intent builder as well. Let's not worry about them now and we can clean this logic up later.